config.py

```diff
@@ -346,6 +347,61 @@ def cmdline(self):
                               'the same ({0})'.format(self.port))
 
     def files(self):
+        '''Read self.PUBLIC_FILE, and for each public address (starting at column 0)
+        construct a dict of the (indented) local addresses following it.
+        '''
+        self.public_servers = dict()
+        try:
+            with open(self.PUBLIC_FILE) as public:
+                errmsg = self.PUBLIC_FILE + ':'
+                self.log(LOG_DEBUG, 'Opened', self.PUBLIC_FILE)
+                phost = ''
+                lineno = 0
+                for line in iter(l.rstrip() for l in public):
+                    lineno += 1
+                    wheremsg = '{0}:{1}:'.format(self.PUBLIC_FILE, lineno)
+                    # ignore blank lines and comments
+                    if not line or line.isspace() or \
+                       line.lstrip().startswith('#'):
+                        continue
+                    # indented lines are local server addresses
+                    if line[0].isspace():
+                        host = line.lstrip()
+                        if not phost:
+                            raise ConfigError(errmsg, 'Error: missing phost address')
+                        # TODO: check if host is valid
+                        if host in self.public_servers[phost].keys():
+                            self.log(LOG_PRINT, wheremsg, 'Warning:',
+                                     host, 'appears multiple times')
+                        self.public_servers[phost][host] = None
+                    # unindented lines start a new phost address
+                    else:
+                        # clean up the old phost address
+                        if phost:
+                            if not self.public_servers[phost]:
+                                raise ConfigError(wheremsg, 'Error: no '
+                                                  'addresses defined for',
+                                                  phost)
+                            else:
+                                # print a message of the form
+                                # public.txt: 'phost': [server1, server2,...]
+                                self.log(LOG_VERBOSE, self.PUBLIC_FILE,
+                                         repr(phost),
+                                         self.public_servers[phost].keys(),
+                                         sep = ': ')
+                        phost = line
+                        # TODO: check if phost is valid
+                        self.public_servers[phost] = dict()
```

Comment:
There is no reason to have a dict of dicts. I see that the featured code is like that; it appears to have been written by monkeys. Just do public_servers[localaddr] = publicaddr